File: mobile/androidApp/src/androidMain/kotlin/com/bissbilanz/android/fasting/FastingSession.kt
package com.bissbilanz.android.fasting

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.time.Duration
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.seconds
import kotlin.time.Instant

/**
 * A single fast — the one currently running ([endedAt] null) or a completed one
 * in the history. Mirrors the iOS `FastingSession`; the running session is the
 * canonical copy and lives in [FastingSessionStore] so the ongoing notification
 * and its End Fast action can read it without going through the repository.
 */
@Serializable
data class FastingSession(
    val id: String,
    val startedAtEpochMs: Long,
    val targetHours: Int,
    val endedAtEpochMs: Long? = null,
) {
    val startedAt: Instant get() = Instant.fromEpochMilliseconds(startedAtEpochMs)

    val endedAt: Instant? get() = endedAtEpochMs?.let { Instant.fromEpochMilliseconds(it) }

    val targetEnd: Instant get() = startedAt + targetHours.hours

    val duration: Duration? get() = endedAt?.let { it - startedAt }

    val reachedTarget: Boolean get() = (duration ?: Duration.ZERO) >= targetHours.hours

    /**
     * Progress toward the target in 0..1. Floors the denominator at a minute so a
     * malformed zero-hour session cannot divide by zero.
     */
    fun progress(now: Instant): Float {
        val target = maxOf(targetHours.hours, 60.seconds)
        return ((now - startedAt) / target).toFloat().coerceIn(0f, 1f)
    }

    fun elapsed(now: Instant): Duration = maxOf(now - startedAt, Duration.ZERO)
}

/**
 * Reads and writes the running fast and the finished-fast history. Backed by
 * plain SharedPreferences rather than the SQLDelight cache: the notification's
 * End Fast receiver needs it too, and the running fast is device-local state.
 * Finished fasts are pushed to the server through the sync queue (see
 * [FastingManager]) so the web history can show them; this store stays the
 * source of truth on the device.
 */
class FastingSessionStore(
    context: Context,
    private val json: Json,
) {
    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    fun loadCurrent(): FastingSession? =
        prefs.getString(KEY_CURRENT, null)?.let {
            runCatching { json.decodeFromString<FastingSession>(it) }.getOrNull()
        }

    fun saveCurrent(session: FastingSession) {
        prefs.edit().putString(KEY_CURRENT, json.encodeToString(session)).apply()
    }

    fun clearCurrent() {
        prefs.edit().remove(KEY_CURRENT).apply()
    }

    /** Finished fasts, most recent first. */
    fun loadHistory(): List<FastingSession> =
        prefs.getString(KEY_HISTORY, null)?.let {
            runCatching { json.decodeFromString<List<FastingSession>>(it) }.getOrNull()
        } ?: emptyList()

    fun appendToHistory(session: FastingSession) {
        saveHistory(listOf(session) + loadHistory().filter { it.id != session.id })
    }

    /** Replaces the history record with the same id; re-sorts by start so an edited start keeps the list in order. */
    fun updateInHistory(session: FastingSession) {
        saveHistory(loadHistory().filter { it.id != session.id } + session)
    }

    fun removeFromHistory(id: String) {
        saveHistory(loadHistory().filter { it.id != id })
    }

    private fun saveHistory(history: List<FastingSession>) {
        val sorted = history.sortedByDescending { it.startedAtEpochMs }.take(HISTORY_LIMIT)
        prefs.edit().putString(KEY_HISTORY, json.encodeToString(sorted)).apply()
    }

    companion object {
        private const val PREFS_NAME = "fasting"
        private const val KEY_CURRENT = "fasting_session_v1"
        private const val KEY_HISTORY = "fasting_history_v1"
        private const val HISTORY_LIMIT = 60
    }
}

File: mobile/iosApp/PhoneShared/Models/FastingSession.swift
import Foundation

/// A single fast — the one currently running (`endedAt == nil`) or a
/// completed one in the history. The canonical copy of the *running* session
/// lives in `FastingSessionStore` (App Group `UserDefaults`), never in
/// SwiftData: the app, the widget extension and the `EndFastIntent`
/// background process all need to read it, and the App Group defaults are
/// the established lightweight cross-process channel (see `WidgetSnapshot`).
struct FastingSession: Codable, Identifiable, Equatable {
    let id: UUID
    var startedAt: Date
    var targetHours: Int
    var endedAt: Date?

    init(id: UUID = UUID(), startedAt: Date, targetHours: Int, endedAt: Date? = nil) {
        self.id = id
        self.startedAt = startedAt
        self.targetHours = targetHours
        self.endedAt = endedAt
    }

    var targetEndDate: Date {
        startedAt.addingTimeInterval(TimeInterval(targetHours) * 3600)
    }

    var duration: TimeInterval? {
        endedAt.map { $0.timeIntervalSince(startedAt) }
    }

    var reachedTarget: Bool {
        (duration ?? 0) >= TimeInterval(targetHours) * 3600
    }

    /// Wire shape for the server copy of a finished fast. The local UUID doubles
    /// as the server id, so a retried upload or an edit that lands before the
    /// first upload drained both hit the same row.
    var upsertBody: FastingSessionUpsert? {
        guard let endedAt else { return nil }
        return FastingSessionUpsert(
            id: id.uuidString.lowercased(),
            startedAt: DateFormatting.isoDateTimeString(from: startedAt),
            endedAt: DateFormatting.isoDateTimeString(from: endedAt),
            targetHours: targetHours
        )
    }

    /// Range driving the date-relative elapsed timers (`Text(timerInterval:)`).
    /// The upper bound is a far-out cap, not the target — an elapsed timer
    /// must keep counting past the target, it only stops at the range's end.
    var elapsedRange: ClosedRange<Date> {
        startedAt ... startedAt.addingTimeInterval(7 * 24 * 3600)
    }

    /// Range driving the date-relative progress views and the remaining
    /// countdown. Guarded so it stays a valid range even for a malformed
    /// zero-hour session.
    var progressRange: ClosedRange<Date> {
        startedAt ... max(targetEndDate, startedAt.addingTimeInterval(60))
    }
}

/// Reads and writes the running fast and the finished-fast history in the
/// shared App Group `UserDefaults`, mirroring `WidgetSnapshotStore`. Both
/// sides degrade gracefully when the suite is unavailable — callers just see
/// "no session".
enum FastingSessionStore {
    static let currentKey = "fasting_session_v1"
    static let historyKey = "fasting_history_v1"
    static let historyLimit = 60

    private static var defaults: UserDefaults? {
        UserDefaults(suiteName: WidgetSnapshotStore.appGroupId)
    }

    static func loadCurrent() -> FastingSession? {
        guard let data = defaults?.data(forKey: currentKey) else { return nil }
        return try? JSONDecoder().decode(FastingSession.self, from: data)
    }

    static func saveCurrent(_ session: FastingSession) {
        guard let defaults, let data = try? JSONEncoder().encode(session) else { return }
        defaults.set(data, forKey: currentKey)
    }

    static func clearCurrent() {
        defaults?.removeObject(forKey: currentKey)
    }

    /// Finished fasts, most recent first.
    static func loadHistory() -> [FastingSession] {
        guard let data = defaults?.data(forKey: historyKey) else { return [] }
        return (try? JSONDecoder().decode([FastingSession].self, from: data)) ?? []
    }

    static func appendToHistory(_ session: FastingSession) {
        saveHistory([session] + loadHistory().filter { $0.id != session.id })
    }

    /// Replaces the record with the same id; re-sorts by start so an edited
    /// start keeps the list in order.
    static func updateInHistory(_ session: FastingSession) {
        saveHistory(loadHistory().filter { $0.id != session.id } + [session])
    }

    static func removeFromHistory(id: UUID) {
        saveHistory(loadHistory().filter { $0.id != id })
    }

    private static func saveHistory(_ history: [FastingSession]) {
        let sorted = Array(history.sorted { $0.startedAt > $1.startedAt }.prefix(historyLimit))
        guard let defaults, let data = try? JSONEncoder().encode(sorted) else { return }
        defaults.set(data, forKey: historyKey)
    }
}
